presto-accumulo/src/main/java/com/facebook/presto/accumulo/index/ColumnCardinalityCache.java

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright 2016 Bloomberg L.P.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.accumulo.index;
+
+import com.facebook.presto.accumulo.conf.AccumuloConfig;
+import com.facebook.presto.accumulo.metadata.AccumuloTable;
+import com.facebook.presto.accumulo.model.AccumuloColumnConstraint;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import org.apache.accumulo.core.client.AccumuloException;
+import org.apache.accumulo.core.client.AccumuloSecurityException;
+import org.apache.accumulo.core.client.BatchScanner;
+import org.apache.accumulo.core.client.Connector;
+import org.apache.accumulo.core.client.TableNotFoundException;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.PartialKey;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.security.Authorizations;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.io.Text;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * This class is an indexing utility to cache the cardinality of a column value for every table.
+ * Each table has its own cache that is independent of every other, and every column also has its
+ * own Guava cache. Use of this utility can have a significant impact for retrieving the cardinality
+ * of many columns, preventing unnecessary accesses to the metrics table in Accumulo for a
+ * cardinality that won't change much.
+ */
+public class ColumnCardinalityCache
```

Comment:
So... Since you're using deprecated Connector API which does not guarantee single-threaded access to `ConnectorMetadata`, you should make sure this class is thread safe. Or move to new Connector API, which would be preferred way anyway. :)